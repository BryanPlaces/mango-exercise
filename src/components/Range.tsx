"use client"

import React from 'react';
import { useSlider, useSliderDrag } from '../hooks';
import InputRange from './InputRange';
import './../styles/styles.scss'

interface RangeProps {
  initialMin?: number;
  initialMax?: number;
  step?: number;
  useDecimals?: boolean;
  allowedValues?: number[];
}

const Range = ({ initialMin, initialMax, step, useDecimals, allowedValues } : RangeProps) => {

  const {
    sliderRef,
    min,
    max,
    minSelectedRange,
    maxSelectedRange,
    leftStylePointer,
    rightStylePointer,
    rangeStyleWidth,
    handleMinChange,
    handleMaxChange,
    handleRightSliderChange,
    handleLeftSliderChange
  } = useSlider({initialMin, initialMax, step, allowedValues});

  const { handleMouseDown: handleMouseDownRight } = useSliderDrag(handleRightSliderChange);
  const { handleMouseDown: handleMouseDownLeft } = useSliderDrag(handleLeftSliderChange);

  const showInputs = allowedValues && allowedValues.length ? false : true;

  return (
    <>
        <div className='range-container'>

          <InputRange id={"minInput"} type={"number"} onlyLabel={showInputs} value={min} onChange={handleMinChange} useDecimals={useDecimals}/>

          <div ref={sliderRef} className='range-wrapper' >
            <div className='slider-pointer' style={{ left: `${leftStylePointer}%` }} onMouseDown={ handleMouseDownLeft } ></div>
            <div className='slider' style={{ left: `${leftStylePointer}%`, width: `${rangeStyleWidth}%` }} ></div>
            <div className='slider-pointer' style={{ left: `${rightStylePointer}%` }} onMouseDown={ handleMouseDownRight } ></div>
          </div>

          <InputRange id={"maxInput"} type={"number"} onlyLabel={showInputs} value={max} onChange={handleMaxChange} useDecimals={useDecimals}/>
        </div>

      <p style={{ textAlign: 'center', marginTop: '10px' }}>Rango: {minSelectedRange} - {maxSelectedRange}</p>
    </>
  );
};

export default Range;