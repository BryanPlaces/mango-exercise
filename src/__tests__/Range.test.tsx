import React from 'react';
import '@testing-library/jest-dom'
import { render } from '@testing-library/react';
import Range from '../components/Range';

describe('Range Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders Range component ussing [min, max]', () => {
    const { container } = render(
      <Range initialMin={0} initialMax={100} step={1} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders Range component ussing the allowedValues', () => {
    const { container } = render(
      <Range allowedValues={[1.99, 5.99, 10.99, 30.99, 50.99, 70.99]} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});