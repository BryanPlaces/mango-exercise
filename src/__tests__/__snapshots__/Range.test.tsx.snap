// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range Component renders Range component ussing [min, max] 1`] = `
<div
  class="range-container"
>
  <label
    class="range-label"
  >
    0
  </label>
  <div
    class="range-wrapper"
  >
    <div
      class="slider-pointer"
      style="left: 0%;"
    />
    <div
      class="slider"
      style="left: 0%; width: 100%;"
    />
    <div
      class="slider-pointer"
      style="left: 100%;"
    />
  </div>
  <label
    class="range-label"
  >
    100
  </label>
</div>
`;

exports[`Range Component renders Range component ussing the allowedValues 1`] = `
<div
  class="range-container"
>
  <label
    class="range-label"
  >
    1.99
  </label>
  <div
    class="range-wrapper"
  >
    <div
      class="slider-pointer"
      style="left: 0%;"
    />
    <div
      class="slider"
      style="left: 0%; width: 100%;"
    />
    <div
      class="slider-pointer"
      style="left: 100%;"
    />
  </div>
  <label
    class="range-label"
  >
    70.99
  </label>
</div>
`;

exports[`Slider Component renders Slider component 1`] = `
<div
  class="range-container"
>
  <label
    class="range-label"
  >
    0
  </label>
  <div
    class="range-wrapper"
  >
    <div
      class="slider-pointer"
      style="left: 0%;"
    />
    <div
      class="slider"
      style="left: 0%; width: 100%;"
    />
    <div
      class="slider-pointer"
      style="left: 100%;"
    />
  </div>
  <label
    class="range-label"
  >
    100
  </label>
</div>
`;
